lib/mongoid_rails_migrations/active_record_ext/migrations.rb

```diff
@@ -268,6 +291,14 @@ def proper_table_name(name)
         name
       end
 
+      def with_mongoid_client(mongoid_client_name, &block)
+        previous_mongoid_client_name = Thread.current[:mongoid_client_name]
+        Thread.current[:mongoid_client_name] = mongoid_client_name
```

Comment:
Using a Thread variable to share and leak the client name is quite intrusive IMO, it works well in our case because we use the same (and with the same value) but it could be very confusing for people using a different variable or sharding implementation (as this gem does not provide the sharding implementation).

I think here we can more simply set the mongo client around each migration without leaking a thread variable, can't we? we don't need this level of precision which is actually more for sharding an application. Here there's only one model and the rest is up to the application.

Reply:
> Using a Thread variable to share and leak the client name is quite intrusive IMO, it works well in our case because we use the same (and with the same value) but it could be very confusing for people using a different variable or sharding implementation (as this gem does not provide the sharding implementation).

Yep but how do we handle the `db:migrate:status`, `db:version` tasks to target a particular client ? Using an environment variable ? That's what I did first but in the end it seemed more harmonious to me to use the same way of switching client in the whole codebase.

> I think here we can more simply set the mongo client around each migration without leaking a thread variable, can't we? we don't need this level of precision which is actually more for sharding an application. Here there's only one model and the rest is up to the application.

Not sure of what you want to say by `around each migration`, the client needs to be changed only when we add the migration entry in the `data_migrations` collection not during the migration execution.

We could replace this method with `Mongoid.with_client` around the code that add/remove entries in `data_migrations` if it's what you asked, yes.

Reply:
> Yep but how do we handle the db:migrate:status, db:version tasks to target a particular client ? Using an environment variable ? That's what I did first but in the end it seemed more harmonious to me to use the same way of switching client in the whole codebase.

What's the problem with these tasks?

> Not sure of what you want to say by around each migration, the client needs to be changed only when we add the migration entry in the data_migrations collection not during the migration execution.
>
> We could replace this method with Mongoid.with_client around the code that add/remove entries in data_migrations if it's what you asked, yes.

I think the standard way to change the mongoid client is with `Mongoid.override_client(name)` now, and we can save/restore the previous one to be clean with something like https://github.com/pavlo/mongoid-dynamic-clients/blob/develop/lib/mongoid/dynamic_clients/client_switch.rb. `with_client` is a monkey-patch from us from way before Mongoid implemented `override_client`.

But yes that's what I mean. And yes theoretically we could only do this for the DataMigration call, but it would be weird and error prone IMO to have a migration with "client :shard" execute with the default client by default don't you think? In our case we don't care I think because our project will change the client anyway but for other users of the gem it may be more logical to have the client designated in the migration during the whole migration execution.